apps/backend/src/services/usersServices.ts

```diff
@@ -0,0 +1,173 @@
+import bcrypt from 'bcrypt';
+import { User } from '@prisma/client';
+import { CreateUserData, UserWithoutId, UserWithoutPasswordHash } from '../types';
+import prismaClient from '../../prismaClient';
+
+const USER_FIELDS_WITHOUT_PASSWORD_HASH = {
+  id: true,
+  user_name: true,
+  first_name: true,
+  last_name: true,
+  dob: true,
+  gender_id: true,
+  bio: true,
+};
+
+// helpers
+
+/**
+ * - narrows username type to string, checks if username is valid
+ * - throws Error if username already exists
+ * - throws Error if username not valid
+ * @param username this should be a valid username
+ * @returns resolved promise with username as value
+ */
+const parseAndValidateUsername = async (username: unknown): Promise<string> => {
+  const MINIMUM_USERNAME_LENGTH = 4;
+  if (typeof username !== 'string' || username.length < MINIMUM_USERNAME_LENGTH) {
+    const error = new Error('Invalid Username');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const existingUser: User | null = await prismaClient.user.findUnique({
+    where: {
+      user_name: username,
+    },
+  });
+
+  if (existingUser) {
+    const error = new Error('User already exists');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  return Promise.resolve(username);
+};
+
+/**
+   * - narrows type of password to string, generates password hash from the password
+   * - throws error if password is missing or invalid ie less than 4 characters
+   * @param password this should be a valid password
+   * @returns Promise resolved to password hash
+   */
+const getPasswordHash = async (password: unknown): Promise<string> => {
+  const MINIMUM_PASSWORD_LENGTH = 4;
+  if (typeof password !== 'string' || password.length < MINIMUM_PASSWORD_LENGTH) {
+    const error = new Error('Invalid Password');
+    error.name = 'UserDataError';
+    throw error;
+  }
+
+  const SALT_ROUNDS = 10;
+  const passwordHash = await bcrypt.hash(password, SALT_ROUNDS);
+  return Promise.resolve(passwordHash);
+};
+
+// user services functions
+
+/**
+ * - performs type narrowing for data from req.body to POST /user
+ * - if username, password present and of type string, then returns an body containing input data
+ * - the mandatory fields can be changed in future
+ * - else throws Error
+ * @param body request.body should contain username and password
+ * @returns Promise resolved to user input to POST /user
+ */
+const parseCreateUserData = async (body: unknown): Promise<CreateUserData> => {
+  if (!body || typeof body !== 'object') {
```

Comment:
I believe lines 78-82 are unnecessary and the type for `body` can be set to `object` in the parameter since it can only ever be an empty object, or contain the properties we want because we've defined `app.use(express.json());`. Please double-check though.

Reply:
I got inspiration for this from FSO. Their logic is that since `req.body` can by anything, `express` sets it to type `any`. Now we have `noExplicitAny` rule hence they suggest to use `unknown`.

But, as you pointed out, we are using `express.json()` so that should automatically convert body to an object. However, if the client does not content-type as `application/json` - then what would `express.json()` do ? Since, we will only ever use the API by ourselves, it should not be an issue.

This is more safe. But if you are sure about it then let me know, we can change it.

Reply:
I'm 99% sure, but let's leave it for now, at least, to be 100%. @radu-constantin @RobertRodes unless you have any further insight on this.